### ISSUE TITLE
SDL 4.2 freezes on quit, ^C generates dump

### ISSUE THREAD

**Issue by pjfarleyiii** [SDL-Hercules-390/hyperion#193]
Tried out the early 4.2 release tonight.  Mostly works (no CCKD64 volumes, only CCKD), but freezes when"quit" command issued by termination script. Ctrl-C is then required to exit, which generates the below attached crash dump.

It also throws two instruction errors at z/OS startup, for the ` STSCH` and `SVS` instructions.  See the below hercules log.

Let me know if there is anything else I can supply to assist.

Peter

__________________________
Crash dump:

_(compressed with 7z-Ultra but extension is ".zip" to get it uploaded)_

* [Hercules-4.2-dmp.zip](https://github.com/SDL-Hercules-390/hyperion/files/2903924/Hercules-4.2-dmp.zip)

&nbsp;
Instruction failures at startup:

```
00:15:30 HHC00801I Processor CP00: Operand exception code 0015  ilc 4
00:15:30 HHC02324I CP00: PSW=0400200080000000 000000002000017C INST=B2342000     STSCH 0(2)                   store_subchannel
00:15:30 HHC02326I CP00: V:0000000020013180:K:06=00000000 00000000 00000000 00000000  ................
00:15:30 HHC02269I CP00: R0=00000000200131BC R1=0000000000050000
00:15:30 HHC02269I CP00: R2=0000000020013180 R3=0000000000000044
00:15:30 HHC02269I CP00: R4=0000000020013180 R5=0000000000000000
00:15:30 HHC02269I CP00: R6=000000002000F250 R7=0000000000000002
00:15:30 HHC02269I CP00: R8=00000000FFFFFFF9 R9=0000000020012000
00:15:30 HHC02269I CP00: RA=0000000000000005 RB=000000002000079C
00:15:30 HHC02269I CP00: RC=00000000A0000000 RD=00000000200006E4
00:15:30 HHC02269I CP00: RE=00000000A0000172 RF=00000000FFFFFFF8
00:15:30 HHC02271I CP00: C0=0000000000000200 C1=0000000000004003
00:15:30 HHC02271I CP00: C2=0000000000000000 C3=0000000000000000
00:15:30 HHC02271I CP00: C4=0000000000000000 C5=0000000000000000
00:15:30 HHC02271I CP00: C6=00000000FE000000 C7=0000000000000000
00:15:30 HHC02271I CP00: C8=0000000000000000 C9=0000000000000000
00:15:30 HHC02271I CP00: CA=0000000000000000 CB=0000000000000000
00:15:30 HHC02271I CP00: CC=0000000000000000 CD=0000000000000000
00:15:30 HHC02271I CP00: CE=00000000C2000000 CF=0000000000000000
00:15:30 HHC00814I Processor CP00: SIGP Unassigned                       (14) CP00, PARM 0000000064000000: CC 1 status 00000002
00:15:32 HHC00801I Processor CP00: Specification exception code 0006  ilc 4
00:15:32 HHC02324I CP00: PSW=0404000080000000 0000000000005176 INST=B2650000     SVS   0,0                    set_vector_summary
00:15:32 HHC02326I CP00: V:0000000000000000:K:0E=000A0000 000130E1 00000000 00000000  ................
00:15:32 HHC02326I CP00: V:0000000000000000:K:0E=000A0000 000130E1 00000000 00000000  ................
00:15:32 HHC02269I CP00: R0=0000000000000000 R1=0000000000000001
00:15:32 HHC02269I CP00: R2=0000000000FDA318 R3=D5BD13D4023EFC20
00:15:32 HHC02269I CP00: R4=0000000000FD9DB8 R5=FFFFFFFF023EFC20
00:15:32 HHC02269I CP00: R6=0000000000FD9DB8 R7=0000000000FD9DB8
00:15:32 HHC02269I CP00: R8=00000000023EFC20 R9=00000000023EFC20
00:15:32 HHC02269I CP00: RA=000000000000517A RB=0000000000000100
00:15:32 HHC02269I CP00: RC=0000000000004D30 RD=0000000000005260
00:15:32 HHC02269I CP00: RE=0000000080004F3E RF=0000000000FD9DB8
00:15:32 HHC02271I CP00: C0=00800002CF88EE60 C1=000000007CF70007
00:15:32 HHC02271I CP00: C2=000000007FEAB840 C3=0000000180000001
00:15:32 HHC02271I CP00: C4=0000000100000001 C5=000000007FEAB900
00:15:32 HHC02271I CP00: C6=00000000FE000000 C7=000000007CF70007
00:15:32 HHC02271I CP00: C8=0000000000000000 C9=0000000000000000
00:15:32 HHC02271I CP00: CA=0000000000000000 CB=0000000000000000
00:15:32 HHC02271I CP00: CC=0000000000000000 CD=000000007CF70007
00:15:32 HHC02271I CP00: CE=00000000C00FF2A4 CF=0000000000000000
```

**Comment by ivan-w** [comment]
I'm going assume the Program Interrupts are a red herring...

I very much suspect the STSCH is perfectly normal (it's NOT a failure)... z/OS s probably scanning all possible subchannels until it hits the wall.. As for the SVS, I'm not sure, but it's probably also testing functionality. 

Program interrupts do NOT always mean there was a failure somewhere !

Now... Quitting herc from scripts has always been a problem...  and especially if the program has registered with the Support Processor to receive termination/shutdown requests.

**Comment by pjfarleyiii** [comment]
Wouldn't using STOPALL and SYSCLEAR before QUIT reset any such registration?   If not, is there any possible solution for shutdown automation?  I have observed that the TK4- shutdown automation always seems to work flawlessly on my machine.  Is that a fluke or maybe the effect of one of the corrections and/or enhancelents in the version of Hyperion delivered with TK4-?  Or just that MVS3.8J never "registers" for such a notification?

Assuming such a notification request has been done, is there any way to tell Hyperion to ignore or clear it?

Peter

**Comment by pjfarleyiii** [comment]
I forgot to ask in my prior comment, do you think that the Spinhawk 3.13 version gets the same interrupts but just doesn't report them as this Hyperion version does? (4.0 version also reports the same ones, as does the TK4- Hyperion version.)

**Comment by Fish-Git** [comment]
> Also throws two instruction errors at z/OS startup, for STSCH and SVS instructions. See pasted lines below from the hercules log.

I've seen the same STSCH and SVS instructions logged on much earlier versions of Hercules, as far back as 4.0.0 (June 2018), so it's nothing new. They _always_ occur with z/OS 2.1 (and possibly z/OS 1.10 too, although I didn't check). So Ivan is right: they're a red herring (completely normal) and _not_ something newly introduced by SDL Hyperion 4.2.

As far as your automated script shutdown problem is concerned, I need to see your script first (please insert it into your GitHub comment reply) so I can try it for myself.

I also need know what guest you are running: z/OS? Which version? z/VM? Which version? I'm presuming you're running z/OS 2.1 based on your complaint about the STSCH and SVS instructions, but I need to know for sure.

Also, are you using a 3270 as your z/OS master console? Or are you letting z/OS default to writing its messages to the HMC? (i.e. to the Hercules panel). That _might_ have an impact on your ability to successfully shutdown Hercules (`quit` or `exit` command) from a Hercules script. I will try it for myself once you show us what your shutdown script looks like.

I have downloaded your dump, but have not analyzed it yet. When I do, I will let you know if I find anything.

**Comment by pjfarleyiii** [comment]
Thanks for the explanation about the instruction failures, I am happy to ignore them.

Guest is z/OS V1.13.  Yes, I use a 3270 master console, but I also automatically at startup via HAO issue ".V CN(*),ACTIVATE" and ".V CN(*),INTIDS=Y,UNKNIDS=Y" at the HMC/Herc console to get z/OS console messages routed to the Herc console for HAO's used much later at shutdown.

Paste of startup and shutdown scripts follows.  Hercules is started in an Admin CMD.EXE window with SET HERCULES_RC=scripts\ipl.rc set before startup.  The automated startup and shutdown is modeled closely on the TK4- model, and uses the exact same directory structure as TK4- does.

The startmst.cmd and starttso.cmd files used as "sh" commands in these scripts each start up a Vista3270 session to port 3270.  The command file stopterm.cmd issues the command "taskkill /IM vistaTN3270.exe" to kill all the Vista sessions in one go.

scripts\ipl.rc

# Complete z/OS HMC activation after IPL
hao tgt HHCCP041I
hao cmd script scripts/cn_activate
hao tgt BPXI004I
hao cmd sh STARTGUI starttso.cmd
hao tgt 01 IEA793A NO DUMP DATA SETS AVAILABLE
hao cmd .1,D
#
# Don't hurry
#
* Start 3270 sessions NOW for Master Console(s)
sh STARTGUI startmst.cmd
pause 5
#
# IPL HDRES1 and kick off z/OS V1R13
#
ipl A80

scripts/cn_activate

# Start Hercules shutdown when OMVS shutdown is complete
#
hao tgt BPXI056E
hao cmd script scripts/pjes2
#
# don't hurry
#
* pausing for 3 seconds, please stand by.
pause 3
#
# Activate the HMC console
#
.V CN(*),ACTIVATE
* pausing for 3 seconds, please stand by.
pause 3
.V CN(*),INTIDS=Y,UNKNIDS=Y

scripts/pjes2

hao tgt HASP085
hao cmd script scripts/z_eod
#
# don't hurry
#
* pausing for 3 seconds, please stand by.
pause 3
#
# Stop JES2
#
.$PJES2

scripts/z_eod

hao tgt IEE985A
hao cmd script scripts/quiesce
#
# don't hurry
#
* pausing for 3 seconds, please stand by.
pause 3
#
# perform end of day processing
#
.Z EOD

scripts/quiesce

hao tgt BLW002I
hao cmd script scripts/poweroff
#
# don't hurry
#
* pausing for 3 seconds, please stand by.
pause 3
#
# Quiesce all CPU's
#
.QUIESCE

scripts/poweroff

* pausing for 3 seconds, please stand by.
pause 3
stopall
* pausing for 3 seconds, please stand by.
pause 3
sysclear
* pausing for 3 seconds, please stand by.
pause 3
sh STARTGUI stopterm.cmd
* pausing for 3 seconds, please stand by.
pause 3
quit

**Comment by pjfarleyiii** [comment]
Apologies for all those large-font and bulleted lines in my prior comment.  It appears that the github reply mechanism recognizes "#' and "*" at start of line as some kind of text markup.  I am packaging up all six scripts in a zip file and uploading the zip file in this reply.

In my prior reply, the vary commands are for "CN(asterisk-character)", it looks like the asterisk got taken as an "italic font" marker as well.

Is there any way to make github treat replies as plain text?

Anyway, regards and thanks for all your work on this.

Peter

[scripts.zip](https://github.com/SDL-Hercules-390/hyperion/files/2912952/scripts.zip)

**Comment by Fish-Git** [comment]
> Paste of startup and shutdown scripts follows further below.

Thank you. I will give your technique a try.

> Is there any way to make github treat replies as plain text?

Sure:

* https://guides.github.com/features/mastering-markdown/
* https://help.github.com/en/articles/basic-writing-and-formatting-syntax

For inline blocks of code I usually just use the [enclose the codeblock within separate triple-backtick lines](https://help.github.com/en/articles/basic-writing-and-formatting-syntax#quoting-code) technique.

For other more language-specific code (such as 'C' code or diffs (patches)) I specify the coding language to cause the code to be nicely syntax highlighted, as you can see in the following comment:

* https://github.com/SDL-Hercules-390/hyperion/issues/188#issuecomment-467604468

Once you start using GitHub's markdown it starts to get easier. But really, don't worry about it. I make it habit of fixing up (editing) other people's comments to make them more readable.

Yes, your original comment looked horrible, due to GitHub treating your scripts as markdown. But as you can see it wasn't too difficult for me to fix it up for you, so don't worry about such things. I'll take care of them for you.&nbsp;`:)`

Thanks for the scripts. I'll give them a try and get back to you.
___
Oh yeah: I've analyzed your crash dump and there seems to be a problem in Hercules's special Ctrl+C handling logic brought about by your doing a Ctrl+C while Hercules was in the middle of shutting down:

* [!analyze -v.txt](https://github.com/SDL-Hercules-390/hyperion/files/2913171/analyze.-v.txt)


I haven't quite figured out the complete dynamics of the bug yet, but once I can recreate it (with, _hopefully_, the help of your scripts) I should _(hopefully!)_ be able to fix it.&nbsp;&nbsp;`<me: fingers crossed>`

**Comment by pjfarleyiii** [comment]
Thanks for the markdown links.  That does seem simple enough, I will keep them in my bookmark list,

I only pressed `Ctrl-C` because it seemed like Hercules was frozen at the `quit` command in `scripts\poweroff`, which is the real bug I guess.

Peter

**Comment by Fish-Git** [comment]
> Thanks for the markdown links.

You're very welcome.

> I only pressed `Ctrl-C` because it seemed like Hercules was frozen at the `quit` command in `scripts\poweroff`, which is the real bug I guess.

I'm not so sure yet! It might be that the `quit` command was hanging, _or,_ it might be that the `quit` command was simply taking an unexpectedly long time to finish (which I guess could be considered a bug in and of itself) and the only bug to be fixed is the crash when `Ctrl+C` is done while Hercules is in the middle of "quitting". It could be either one. Or both. I haven't had a chance to try and repro your issue yet. I'm currently busy working on another (higher priority IMO) issue: #194.

Once I'm done with that issue (or reach a stalling point) I'll switch over to working on your issue.**(*)**
___
**(*)**: _I work best when I concentrate on only one thing at a time. When I try working on too many different things simultaneously I only end up screwing one or the other or both of them up. My brain isn't very good at multi-tasking highly technical issues. I've always been that way._

**Comment by pjfarleyiii** [comment]
Understood and agree my issue is not a priority.  I'll check back as time permits (RL work busy for the next few days anyway), at least daily.

And I'm the same way about my time management.  One complex technical issue at a time, thank you.

Happy debugging.

Peter

**Comment by Fish-Git** [comment]
> _**EDIT:**_  [...] The target should be:
> 
> ```
> hao tgt HHC00006I
> ```
> `HHCCP041I` is specific to Spinhawk V3.13.

Thanks. Fixed.

**Comment by Fish-Git** [comment]
Peter (@pjfarleyiii),

> The `startmst.cmd` and `starttso.cmd` files used as `sh` commands in these scripts each start up a Vista3270 session to port 3270.

I very much need to see these two scripts, because I think maybe that might be where your problem lies.

**Comment by pjfarleyiii** [comment]
As requested:

startmst.cmd

```
@Echo off
start "zpdt Console" /D%USERPROFILE%\Documents\TomBrennanSoftware\VistaTN3270 "%ProgramFiles(x86)%\TomBrennanSoftware\VistaTN3270\vistaTN3270.exe" zpdtcons.ses
```

starttso.cmd

```
@Echo off
start "zpdt TSO" /D%USERPROFILE%\Documents\TomBrennanSoftware\VistaTN3270 "%ProgramFiles(x86)%\TomBrennanSoftware\VistaTN3270\vistaTN3270.exe" zpdt.ses
```

**Comment by Fish-Git** [comment]
> > > The `startmst.cmd` and `starttso.cmd` files used as `sh` commands in these scripts each start up a Vista3270 session
> > > to port 3270.
> > 
> > I very much need to see these two scripts, because I think maybe
> > that might be where your problem lies.
> 
> As requested:
> 
> startmst.cmd
> [...]
> 
> starttso.cmd
> [...]

Thank you. I gave your scripts a try but unfortunately I was unable to reproduce your hang. The z/OS test guest I use came up just fine (fully automated! Cool!), and, after entering the `s shutdown` command to the guest, performed a fully automated shutdown and Hercules exit as well (very slick!). Hercules did not hang however. It exited just fine.

So now I'm left with just the crash that occured when you entered Ctrl+C during shutdown but before it finished. I have not tried to recreate that yet, but that is my next task. The hang however, there is nothing more I can do about since I am unfortunately unable to recreate it.

I don't suppose this hang issue is something that you are able to reliably reproduce at will, is it? I got the impression from your problem report that this was a one time occurrence. Is that correct?

(p.s. I like your scripts! Very slick!)

**Comment by pjfarleyiii** [comment]
I only tried SDL 4.2 the one time, so I don't yet know if it is reproducible.  I will set up the test again and let you know if I can reproduce the hang.

Peter

P.S. - Thanks, but I must give credit where credit is due - Jurgen's TK4- is a superbly crafted piece of work.  I just built on what he created, adjusted for a newer OS.

**Comment by pjfarleyiii** [comment]
I ran several startup/shuitdowns using SDL 4.2 in a row, some with TSO and job activity and some not, but none of them produced a hang at termination. At this point I have to say I think the hang at termination is not yet reproducible on my machine.

I guess the only option at this point is for you to shoot the Ctrl-C bug you saw in the dump and relegate the hang to another day's issue when it can be reliably reproduced.  I think we may be able to test any fix you craft by commenting out the "quit" command in the poweroff script and entering Ctrl-C when poweroff completes without quitting.  It's not *quite* the same conditions, but it's at least close.

I will try some 3390-54 CCKD64 ICKDSF runs next to see if I see any of the symptoms of the other issue you were working on.  ATM I have (I think) 10 initiators available by default, so I can run up to 10 ICKDSF's at a time and see what happens.

Regards,

Peter

**Comment by Fish-Git** [comment]
> > (p.s. I like your scripts! Very slick!)
> 
> Thanks, but I must give credit where credit is due - Jurgen's TK4- is a superbly crafted piece of work. I just built on what he created, adjusted for a newer OS.

Jürgen in indeed amazing and I am honored he is a member of the SDL Hyperion development team.

> I ran several startup/shutdowns using SDL 4.2 in a row, some with TSO and job activity, and some not, but _none_ of them produced a hang at termination. At this point I have to say I think the hang at termination is not yet reproducible on my machine.

Thanks for the confirmation that your original hang was likely a fluke. Let's not waste any more time on it and focus instead on the crash. _That_ should not be happening!

> I guess the only option at this point is for you to shoot the Ctrl-C bug...

Which has already been shot and killed. The fix will be commited shortly.

> I will try some 3390-54 CCKD64 ICKDSF runs next to see if I see any of the symptoms of the other issue you were working on. ATM I have (I think) 10 initiators available by default, so I can run up to 10 ICKDSF's at a time and see what happens.

Thanks. I'm very anxious (keen) on ensuring a solid CCKD64 release. It's an important new feature and I want to make sure there isn't anything I missed or screwed up somewhere. I've been using it myself _exclusively_ on all of my Hercules test systems ever since I did the initial commit for it (I always [eat my own dog food](https://en.wikipedia.org/wiki/Eating_your_own_dog_food)), and have personally not had a lick of trouble with it.

Of course, none of my Hercules test systems actually do much either. I don't really have anything in place to give any of them "a good workout". I don't really stress any of them that much, so I guess that's not really saying much.

_You our users,_ on the other hand, are more likely to give Hercules a good workout by actually using it to do useful things, which is why I'm so interested in getting as many of you as possible to help me test it out. You guys are more likely to do a better job at testing it than the quick/simple IPL, do a few network things, submit a job or two, and shutdown type of tests that I typically do.

So yeah, I'd really appreciate you trying to give it a good workout, Peter. Thanks.

I'll post the details of the Ctrl+C bug (just for the record and for the benefit of my fellow developers) once I commit the fix. If you're setup to build Hercules yourself, just do a `git pull` when you see my commit and then rebuild to pick it up.

If you're _not_ setup to build Herc for yourself and are instead using the pre-built binaries, then I'm afraid that until 4.2 is officially released, you're just going to have to be very careful to _not press Ctrl+C or close the window!!_&nbsp;`;-)`

**Comment by Fish-Git** [comment]
Fixed by commit 09e5007cbe8b2c926ef8e6315a21b8735bc7bfeb.

As promised, here are the details of the bug:

The root of the problem was that the faulting thread (0000000000000e64) was _not_ a Hercules created thread!  It was a _Windows internally created thread_ that Windows created when the console process was first created and whose sole purpose is to perform signal handling (i.e. to handle Ctrl+C, Close, Ctrl+Break, Logff, and Shutdown events).

As a result, the `FindFTHREAD()` function (that the `fthread_getschedparam()` function called to retrieve the pointer to its internal threads list entry) failed, returning a NULL pointer, leading to the crash.  The second exception parameter (Parameter[1]: 0000000000000028) provides evidence of this: 0x28 is decimal 40 which is the displacement of the `nPriority` field our internal `FTHREAD` structure!

The sequence of events was as follows:

  1. User presses Ctrl+C (or clicks the [x] Close box to close the window).

  2. Windows internally handles the event, passing control to its internal thread that handles such signals.

  3. Windows calls our Hercules handler because that's what we asked it to do. (i.e. the `impl.c` call to `SetConsoleCtrlHandler()` on line 1101).

  4. Our `console_ctrl_handler()` function calls `do_emergency_shutdown()`.
  
  5. `do_emergency_shutdown()` calls `do_shutdown()` (with the `sysblk.shutimmed` flag set to true).
  
  6. `do_shutdown()` calls `do_shutdown_now()`.
  
  7. `do_shutdown_now()` calls `hdl_atexit()`.
  
  8. `hdl_atexit()` begins calling all of the registered at-exit functions, one of the last of which is `logger_term()`.
     
  9. The `logger_term()` function attempts to issue the `HHC00101I` thread termination message, one of the parameters of which is the thread's priority.

  10. The `get_thread_priority()` macro calls the `hthread_get_thread_prio()` function.

  11. `hthread_get_thread_prio()` calls `hthread_getschedparam()` (which [the WinDbg analysis](https://github.com/SDL-Hercules-390/hyperion/files/2913171/analyze.-v.txt) reports is where our crash occurred) which gets routed directly to the fthread function `fthread_getschedparam()`. This is why I originally missed seeing the bug: the crash didn't actually occur in the hthread_getschedparam() function, but rather in the _"**fthread_getschedparam()**" function!_ Compiler optimzation was hiding that fact from me! (I _should_ have been smarter though, and should have looked at the function call that that statement was making. If I would have done that I would have caught this bug sooner.)

  12. `fthread_getschedparam()` calls its internal `FindFTHREAD()` function, passing it the thread-id of the current thread (which, as explained, is _not_ an fthread thread!). So `FindFTHREAD()` ends up returning a NULL pointer ("not found"), which, as soon as it is used, causes Hercules to immediately crash with a NULL pointer dereference (i.e. an access violation, invalid pointer read): _**Ka-BOOM!**_

So this was a simple(?) case of a careless bug in fthreads. It shouldn't have been presuming it would only be called from its own threads.  Due to the way that Microsoft designed their Windows signal handling, Hercules code can actually be called by a non-Hercules thread via the `console_ctrl_handler()` signal handling function (which Windows calls _from its own internal thread_ in response to the user pressing Ctrl+C or clicking [x] close, etc).

The fix of course was simple: correct the bad code in `fthread_getschedparam()` that was erroneously presuming its call to `FindFTHREAD()` would always succeed, and deal with the possible failure accordingly.

It's been a long time since fthreads has ever failed (caused problems), but I guess even _I'm_ not perfect _all_ of the time!&nbsp;`>;-)` &nbsp;&nbsp;&nbsp;&lt;jk!&gt;

In the mean time, until 4.2 can be officially released, _**DON'T PRESS CTRL+C OR CLOSE THE WINDOW WHILE HERCULES IS STILL RUNNING!!**_&nbsp;`%-P`

**Comment by pjfarleyiii** [comment]
I created 10 CCKD64 volumes with this (model) command:

```
dasdinit64 -bz2 -lfs USR600.cckd64 3390-54 USR600
```
Which created 10 dasd files with names like USR600.cckd64, USR601.cckd64, etc. up to USR609.cckd64.

Then I added the 10 new DASD devices to the config file with (model) lines like this:

```
0AB3    3390    dasd\USR600.cckd64 ro sf=shadow/zaus60_
```

Then I IPL'ed z/OS and varied the 10 new units offline.  Then I ran 10 jobs like the following model job all at the same time, varying the job name last character from 0 to 9, the volume names from 0 to 9, and the UNIT parameters as per the new config units:

```
//IBMUSER0 JOB (ACCT),CLASS=A,MSGCLASS=H,NOTIFY=&SYSUID   
//FORMAT   EXEC PGM=ICKDSF,REGION=0M,PARM='NOREPLYU,FORCE'
//SYSPRINT DD  SYSOUT=*                                   
//SYSIN    DD  *                                          
  INIT UNIT(AB3) NVAL NVFY VOLID(USR600) PURGE -          
    VTOC(0,1,2939) INDEX(196,0)                           
  TRKFMT UNIT(AB3) NVFY CYLR(206,65519) ERASE             
//*                                                       
```
The 10 jobs all ran together for about the same elapsed time (148+ minutes) and all 10 completed successfully.  No Hercules crash, and shutdown completed normally.

I ran this test with 4 emulated CP engines with 8G Hercules memory and it pegged all four CP's at 100% while the jobs were running on my Win10 Core i7-6700 4.00GHz machine.  The PC ran at slightly over 70% CPU utilization for the duration of the test.

Let me know if there is any variation of this test you would like to see run.

HTH

Peter
